okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java

```diff
@@ -982,144 +752,186 @@ void listModulesFromInterface(String tenantId,
       if (options.getSimulate()) {
-        fut.handle(new Success<>(tml));
-      } else {
-        installAutodeploy(t, pc, options, modsAvailable, tml, tml.iterator(),
-            res1 -> {
-              if (res1.failed()) {
-                fut.handle(new Failure<>(res1.getType(), res1.cause()));
-              } else {
-                fut.handle(new Success<>(tml));
+        return Future.succeededFuture(tml);
+      }
+      return jobs.add(t.getId(), job.getId(), job).compose(res2 -> {
+        Promise<List<TenantModuleDescriptor>> promise = Promise.promise();
+        Future<Void> future = Future.succeededFuture();
+        if (options.getAsync()) {
+          List<TenantModuleDescriptor> tml2 = new LinkedList<>();
+          for (TenantModuleDescriptor tm : tml) {
+            tml2.add(tm.cloneWithoutStatus());
+          }
+          promise.complete(tml2);
+          future = future.compose(x -> {
+            for (TenantModuleDescriptor tm : tml) {
+              tm.setStage(TenantModuleDescriptor.Stage.pending);
+            }
+            return jobs.put(t.getId(), job.getId(), job);
+          });
+        }
+        if (options.getDeploy()) {
+          if (options.getAsync()) {
+            future = future.compose(x -> {
+              for (TenantModuleDescriptor tm : tml) {
+                tm.setStage(TenantModuleDescriptor.Stage.deploy);
               }
+              return jobs.put(t.getId(), job.getId(), job);
             });
-      }
+          }
+          future = future.compose(x -> autoDeploy(t, modsAvailable, tml));
+        }
+        for (TenantModuleDescriptor tm : tml) {
+          if (options.getAsync()) {
+            future = future.compose(x -> {
+              tm.setStage(TenantModuleDescriptor.Stage.call);
+              return jobs.put(t.getId(), job.getId(), job);
+            });
+          }
+          if (options.getIgnoreErrors()) {
+            Promise<Void> promise1 = Promise.promise();
+            installTenantModule(t, pc, options, modsAvailable, tm).onComplete(x -> {
+              if (x.failed()) {
+                logger.warn("Ignoring error for tenant {} module {}",
+                    t.getId(), tm.getId(), x.cause());
+              }
+              promise1.complete();
+            });
+            future = future.compose(x -> promise1.future());
+          } else {
+            future = future.compose(x -> installTenantModule(t, pc, options, modsAvailable, tm));
+          }
+          if (options.getAsync()) {
+            future = future.compose(x -> {
+              if (tm.getMessage() == null) {
+                tm.setStage(TenantModuleDescriptor.Stage.done);
+              }
+              return jobs.put(t.getId(), job.getId(), job);
+            });
+          }
+        }
+        if (options.getDeploy()) {
+          future.compose(x -> autoUndeploy(t, modsAvailable, tml));
+        }
```

Comment:
Should we check async here and update tml stage to `undeploy`? BTW `stage.undeploy` has not been used anywhere in the project yet.

Reply:
Yes. I have change things so that deploy and undeploy only change state for particular modules (which might be a subset of all modules in the list).